### ISSUE TITLE
[NFR]: SQL Instrumental key pagination

### ISSUE THREAD

**Issue by TimurFlush** [phalcon/cphalcon#14754]
Hello there!

My feature request is to refactoring the Phalcon\Paginator\Adapter\QueryBuilder class.

To increase productivity, you need to change the current offset pagination to instrumental pagination on the keys.

A detailed article can be read at https://use-the-index-luke.com/no-offset

If possible, why not implement this feature in our favorite framework that offers so many benefits?

**Comment by niden** [comment]
Closing in favor of https://github.com/phalcon/cphalcon/issues/14608

Please vote for this feature there!

**Comment by ruudboon** [comment]
Great suggestion! I would suggest to not fully refactor the pagination but add this as an option. A page number list can be required for projects. That way we can introduce it in a minor (4.1/4.2) version as well.

**Comment by TimurFlush** [comment]
@ruudboon That's what I meant, but my laziness was just a replacement (:

**Comment by ruudboon** [comment]
@TimurFlush For a project I'm working with a big table that is using pagination. Currently experimenting with this option.
```
 SELECT id, name, address, phone
FROM customers
INNER JOIN (
SELECT id
FROM customers
ORDER BY name
LIMIT 10 OFFSET 990)
AS my_results USING(id);
```
See: 2. Try using a Deferred Join
https://www.iheavy.com/2013/06/19/3-ways-to-optimize-for-paging-in-mysql/

Maybe it can help you as well.

**Comment by niden** [comment]
Resolved in #17021 

Thank you @TimurFlush
